packages/react/src/utils/mergeThemes.ts

```diff
@@ -144,6 +144,32 @@ export const mergeComponentStyles__DEV = (
   }, initial)
 }
 
+const mergeComponentStylesCache = new WeakMap()
+
+export const mergeComponentStylesWithCache = (
+  theme,
+  hashObj,
+  sources: (ComponentSlotStylesInput | null | undefined)[],
+) => {
+  try {
+    if (!mergeComponentStylesCache.get(theme)) {
+      mergeComponentStylesCache.set(theme, {})
+    }
+
+    const hashVal = JSON.stringify(hashObj)
```

Comment:
@levithomason I am using now WeakMap for the theme object, although I still have to use some hash for the props object, as it is different reference always, so I am keeping the `JSON.stringify` approach. I will re-run again on my machine all perf tests and see how this is different than master. Let me know if you have some other suggestions regarding the hashing.